Server/src/main/java/org/gluu/oxauth/token/ws/rs/TokenRestWebServiceImpl.java

```diff
@@ -490,6 +495,103 @@ public Response requestAccessToken(String grantType, String code,
         return response(builder, oAuth2AuditLog);
     }
 
+    /**
+     * Processes token request for device code grant type.
+     * @param grantType Grant type used, should be device code.
+     * @param client Client in process.
+     * @param deviceCode Device code generated in device authn request.
+     * @param scope Scope registered in device authn request.
+     * @param request HttpServletRequest
+     * @param response HttpServletResponse
+     * @param oAuth2AuditLog OAuth2AuditLog
+     */
+    private Response processDeviceCodeGrantType(final GrantType grantType, final Client client, final String deviceCode,
+                                                String scope, final HttpServletRequest request,
+                                                final HttpServletResponse response, final OAuth2AuditLog oAuth2AuditLog) {
+        if (!TokenParamsValidator.validateGrantType(grantType, client.getGrantTypes(), appConfiguration.getGrantTypesSupported())) {
+            return response(error(400, TokenErrorResponseType.INVALID_GRANT, "Grant types are invalid."), oAuth2AuditLog);
+        }
+
+        log.debug("Attempting to find authorizationGrant by deviceCode: '{}'", deviceCode);
+        final DeviceCodeGrant deviceCodeGrant = authorizationGrantList.getDeviceCodeGrant(deviceCode);
+
+        log.trace("DeviceCodeGrant : '{}'", deviceCodeGrant);
+
+        if (deviceCodeGrant != null) {
+            if (!deviceCodeGrant.getClientId().equals(client.getClientId())) {
+                return response(error(400, TokenErrorResponseType.INVALID_GRANT, "The client is not authorized."), oAuth2AuditLog);
+            }
+            if (!deviceCodeGrant.isTokensDelivered()) {
+                RefreshToken refToken = deviceCodeGrant.createRefreshToken();
+                log.debug("Issuing refresh token: {}", refToken.getCode());
+
+                AccessToken accessToken = deviceCodeGrant.createAccessToken(request.getHeader("X-ClientCert"), new ExecutionContext(request, response));
+                log.debug("Issuing access token: {}", accessToken.getCode());
+
+                IdToken idToken = deviceCodeGrant.createIdToken(
+                        null, null, accessToken, refToken,
+                        null, deviceCodeGrant, false, null);
+
+                deviceCodeGrant.setTokensDelivered(true);
+                deviceCodeGrant.save();
```

Comment:
Once tokens are issued deviceCode has to be removed similar to AuthorizationCode.